### PR TITLE
[SYCL] Make vec load store member functions templated

### PR DESCRIPTION
**Pull request by alexbatashev**
The SYCL specification mentions load and store member functions as
```cpp
template <access::address_space addressSpace>
void load(size_t offset, multi_ptr<const dataT, addressSpace> ptr);
template <access::address_space addressSpace>
void store(size_t offset, multi_ptr<dataT, addressSpace> ptr) const;
```
This patch aligns our implementation with the definition that is provided by the spec.

Signed-off-by: Alexander Batashev <alexander.batashev@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/types.hpp [hunk 1]
```diff
@@ -1548,28 +1556,15 @@ class SwizzleOp {
...
   // Leave store() interface to automatic conversion to vec<>.
   // Load to vec_t and then assign to swizzle.
-#ifdef __SYCL_LOAD
-#error "Undefine __SYCL_LOAD macro"
-#endif
-#define __SYCL_LOAD(Space)                                                     \
-  void load(size_t offset, multi_ptr<DataT, Space> ptr) {                      \
+  template <access::address_space Space>
+  void load(size_t offset, multi_ptr<DataT, Space> ptr) {
     vec_t Tmp;                                                                 \
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/types.hpp]:
```suggestion
    vec_t Tmp;
```

File: sycl/include/CL/sycl/types.hpp [hunk 2]
```diff
@@ -1548,28 +1556,15 @@ class SwizzleOp {
...
     Tmp.template load(offset, ptr);                                            \
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/types.hpp]:
```suggestion
    Tmp.template load(offset, ptr);
```

File: sycl/include/CL/sycl/types.hpp [hunk 3]
```diff
@@ -644,30 +644,38 @@ template <typename Type, int NumElements> class vec {
...
+    multi_ptr<const DataT, Space> ConstPtr(Ptr);
+    load(Offset, ConstPtr);
+  }
+  template <int Dimensions, access::mode Mode,
+            access::placeholder IsPlaceholder, access::target Target>
+  void load(size_t Offset,
+            accessor<DataT, Dimensions, Mode, Target, IsPlaceholder> Acc) {
+    multi_ptr<const DataT, detail::TargetToAS<Target>::AS> MultiPtr(Acc);
+    load(Offset, MultiPtr);
+  }
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/types.hpp]: Cannot find such overload in the spec

File: sycl/include/CL/sycl/types.hpp [hunk 4]
```diff
@@ -644,30 +644,38 @@ template <typename Type, int NumElements> class vec {
...
+    for (int I = 0; I < NumElements; I++) {
+      *multi_ptr<DataT, Space>(Ptr + Offset * NumElements + I) = getValue(I);
+    }
+  }
+  template <int Dimensions, access::mode Mode,
+            access::placeholder IsPlaceholder, access::target Target>
+  void store(size_t Offset,
+             accessor<DataT, Dimensions, Mode, Target, IsPlaceholder> Acc) {
+    multi_ptr<DataT, detail::TargetToAS<Target>::AS> MultiPtr(Acc);
+    store(Offset, MultiPtr);
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/types.hpp]: Same here

File: sycl/include/CL/sycl/types.hpp (hunk 3 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/types.hpp]: @AlexeySachkov but it is being used in CTS

Comment (AlexeySachkov) [reply sycl/include/CL/sycl/types.hpp]: @alexbatashev, then it is either CTS bug or spec bug.

Tagging @bader, @mkinsner and @jbrodman for visibility

Comment (alexbatashev) [reply sycl/include/CL/sycl/types.hpp]: @AlexeySachkov I submitted an issue to track this question https://github.com/KhronosGroup/SYCL-CTS/issues/36

**Pull request merged**